modules/apps/roles/roles-admin-web/src/main/java/com/liferay/roles/admin/web/internal/group/type/contributor/util/GroupTypeContributorUtil.java

```diff
@@ -50,24 +51,17 @@ public static long[] getClassNameIds() {
 		return ListUtil.toLongArray(classNameIds, Long::valueOf);
 	}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY
-	)
-	protected void addGroupTypeContributor(
-		GroupTypeContributor groupTypeContributor) {
-
-		_groupTypeContributors.add(groupTypeContributor);
+	@Activate
+	protected void activate(BundleContext bundleContext) {
```

Comment:
@JanisZhang Basically we should avoid such component declaration in util classes.

One way is to make it a real component, and pass the service instance to the JSP.

For example, currently this is used by 3 display context classes which are instantiated in 3 different JSP files. You can try to check whether these JSPs have a common parent in the render chain (is it `RolesAdminPortlet`?). If so, you can put the registry instance as a request attribute, and retrieve it in the JSPs.